src/dev/flang/be/jvm/JVM.java

```diff
@@ -1002,7 +1002,30 @@ else if (r == -1)   // a constructor
         return
           traceReturn(cl, pre)
           .andThen(getf)
-          .andThen(ft.return0());
+          .andThen(
+            fieldExists(r)
+              ? ft.return0()
+              /**
+               * Example where fieldExists is false but we still need a return:
+               *
+               * unit_like : choice unit is
+               * test0(T type, a T) =>
+               *   concur
+               *     .atomic a
+               *     .read
+               * _ := test0 unit_like unit
+               */
+              : _fuir.clazzIsChoice(t)
+              ? Expr.RETURN
```

Comment:
Good to have the comment with the example. 
In this case, I think the condition should better be
```
  ft == PrimitiveType.type_void
  ? ft.return0()
```
such that this also covers non-choice unit types. 

Then the code is the same as in the first case, so we could have
```
          .andThen(
            fieldExists(r) ||
              /**
               * Example where fieldExists is false but we still need a return:
               *
               * unit_like : choice unit is
               * test0(T type, a T) =>
               *   concur
               *     .atomic a
               *     .read
               * _ := test0 unit_like unit
               */
            ft == PrimitiveType.type_void
              ? Expr.RETURN
```